django/db/__init__.py

```diff
@@ -36,6 +37,50 @@
 ]
 
 connections = ConnectionHandler()
+async_connections = AsyncConnectionHandler()
+
+
+class new_connection:
+    """
+    Asynchronous context manager to instantiate new async connections.
+
+    """
+
+    def __init__(self, using=DEFAULT_DB_ALIAS):
+        self.using = using
+
+    async def __aenter__(self):
+        conn = connections.create_connection(self.using)
+        if conn.supports_async is False:
+            raise NotSupportedError(
+                "The database backend does not support asynchronous execution."
+            )
+
+        self.force_rollback = False
+        if async_connections.empty is True:
+            if async_connections._from_testcase is True:
+                self.force_rollback = True
```

Comment:
is this behavior emulating something in the existing connection code?

My read of this code is:
- if there are no connections right now _and_ we're in a testcase, then we'll disable autocommit (i.e. start a transction
- later on, in `__aexit__`, if the connection is not in autocommit, we'll commit the results, except for when this is a test case

I _think_ we should get rid of the `_from_testcase` logic here and get it into `SimpleTestCase`.

One thing I did in my branch is the following:

```
        # in _setup_and_call
        # Convert async test methods.
        if iscoroutinefunction(testMethod):
            setattr(
                self,
                self._testMethodName,
                # note the usage of connect_db_then_run
                async_to_sync(self.connect_db_then_run(testMethod)),
            )

    # added method on SimpleTestCase
    def connect_db_then_run(self, test_method):

        import functools
        from contextlib import AsyncExitStack
        from django.db import new_connection

        @functools.wraps(test_method)
        async def cdb_then_run(*args, **kwargs):
            async with AsyncExitStack() as stack:
                # connect to all the DBs
                for db in self.databases:
                    await stack.enter_async_context(new_connection(using=db))
                await test_method(*args, **kwargs)

        return cdb_then_run
```

Maybe something similar here can avoid the `_from_testcase` magic here (might require some new parameters for `new_connection.__init__` though)

Reply:
The purpose of the attribute was to emulate automatic rollback of tests, like [we do for sync connections](https://github.com/django/django/blob/main/django/test/testcases.py#L1380-L1384).

I agree it would be better to have this feature contained in `testcases.py` and not leak into here.

One issue is that I don't want async tests to start a connection automatically. The connection would be created inside the test method, so there's no way to access it from the outer scope of `SimpleTestCase`.

One option could be to add a new `force_rollback` parameter to `new_connection`, and have `SimpleTestCase` monkey-patch it to pass it as `True`

Reply:
the transaction-based rollback happens for `TestCase`, and not for `SimpleTestCase`/`TransactionTestCase`, so we would want to emulate that behavior I think. 

I don't agree with requiring test writers to open an async connection manually over and over, but I think it's not trivial to avoid that in the current test suite (very much would not mind using pytest...), and current async tests are not using async connections, so I'm not going to insist on automatic connections.

How about the following:
- We add `force_rollback` as a parameter to `new_connection`
- We add a check in `new_connection` that raises an error if we're in `TestCase` and don't pass in `force_rollback=True` (check only happens if the `CHECK_ASYNC_CONNECTION_ROLLBACK_IN_TESTS` setting is set) (but don't do this for `SimpleTestCase`/`TransactionTestCase` on account of those having DB cleanup happening via truncation)

I do not even believe we need the second option just yet, but it feels like the least invasive version of the check, even though it makes `new_connection` calls in tests verbose for now (it's at least solvable by a helper method for people writing many of those tests).

Reply:
After working on some more tests... I now think that explicit connection opening makes sense right now. Still wish we had a _bit_ of a nicer answer to the rollback but we'll get there. But yeah at the very least if I can pass in `force_rollback` as a parameter then I'll be happy here